File: packages/backstage-ts/src/broadcast.ts
/**
 * Backstage SDK - Broadcast
 */

import {
  BROADCAST_STREAM,
  type StreamMessage,
  type RedisClient,
  parseFields,
} from './types';
import { Logger, createLogger, LogLevel, type LoggerConfig } from './logger';
import type { Worker } from './worker';

export interface BroadcastConfig {
  /** Worker instance to extract redis and workerId from */
  worker?: Worker;
  /** Redis client (required if worker not provided) */
  redis?: RedisClient;
  /** Worker ID (required if worker not provided) */
  workerId?: string;
  /** Idle threshold for consumer cleanup in milliseconds (default: 1 hour) */
  consumerIdleThreshold?: number;
  /** Logger configuration */
  loggerConfig?: LoggerConfig;
}

/**
 * Handles broadcasting messages to all active workers.
 * Uses a dedicated broadcast stream with per-worker consumer groups.
 *
 * @example
 * ```typescript
 * const broadcast = new Broadcast({ worker });
 * await broadcast.initialize();
 * await broadcast.send('config-update', { version: 2 });
 * ```
 */
export class Broadcast {
  private redis: RedisClient;
  private workerId: string;
  private consumerGroup: string;
  private logger: Logger;
  private consumerIdleThreshold: number;

  /**
   * Create a new Broadcast instance.
   *
   * @param config - Configuration options
   */
  constructor(config: BroadcastConfig) {
    // Extract redis and workerId from worker or use provided values
    if (config.worker) {
      this.redis = config.worker.redis;
      this.workerId = config.worker.workerId;
    } else if (config.redis && config.workerId) {
      this.redis = config.redis;
      this.workerId = config.workerId;
    } else {
      throw new Error(
        'Broadcast requires either a worker instance or both redis and workerId',
      );
    }

    this.consumerGroup = `broadcast-${this.workerId}`;
    this.consumerIdleThreshold = config.consumerIdleThreshold ?? 60 * 60 * 1000; // 1 hour
    this.logger = createLogger({
      level: LogLevel.INFO,
      ...config.loggerConfig,
    });
  }

  /**
   * Initialize the broadcast system.
   * Creates a unique consumer group for this worker on the broadcast stream.
   *
   * @returns Promise that resolves when initialized
   */
  async initialize(): Promise<void> {
    try {
      await this.redis.send('XGROUP', [
        'CREATE',
        BROADCAST_STREAM,
        this.consumerGroup,
        '0',
        'MKSTREAM',
      ]);
      this.logger.debug(`Created consumer group: ${this.consumerGroup}`);
    } catch (err: unknown) {
      if (err instanceof Error && !err.message.includes('BUSYGROUP')) {
        throw err;
      }
    }
  }

  /**
   * Send a broadcast message to all listening workers.
   *
   * @param taskName - Name of the broadcast task
   * @param payload - Data payload
   * @returns The message ID
   */
  async send(taskName: string, payload: unknown): Promise<string> {
    const messageId = await this.redis.send('XADD', [
      BROADCAST_STREAM,
      '*',
      'taskName',
      taskName,
      'payload',
      JSON.stringify(payload),
      'enqueuedAt',
      String(Date.now()),
    ]);

    return messageId as string;
  }

  /**
   * Read new broadcast messages for this worker.
   *
   * @param blockMs - Milliseconds to block if no messages available (default 0)
   * @returns Array of stream messages
   */
  async read(blockMs: number = 0): Promise<StreamMessage[]> {
    const messages: StreamMessage[] = [];

    try {
      const result = await this.redis.send('XREADGROUP', [
        'GROUP',
        this.consumerGroup,
        this.workerId,
        'COUNT',
        '10',
        'BLOCK',
        String(blockMs),
        'STREAMS',
        BROADCAST_STREAM,
        '>',
      ]);

      if (!result || !Array.isArray(result) || result.length === 0) {
        return messages;
      }

      for (const streamEntry of result) {
        if (!Array.isArray(streamEntry) || streamEntry.length < 2) continue;

        const [, streamMessages] = streamEntry as [string, unknown[]];
        if (!Array.isArray(streamMessages)) continue;

        for (const msgEntry of streamMessages) {
          if (!Array.isArray(msgEntry) || msgEntry.length < 2) continue;

          const [msgId, fields] = msgEntry as [string, unknown[]];
          const message = this.parseMessage(msgId, fields);
          if (message) {
            messages.push(message);
          }
        }
      }
    } catch {
      // Skip read errors
    }

    return messages;
  }

  /**
   * Acknowledge a broadcast message.
   * This marks it as processed for this worker's consumer group.
   *
   * @param messageId - ID of the message to acknowledge
   */
  async ack(messageId: string): Promise<void> {
    await this.redis.send('XACK', [
      BROADCAST_STREAM,
      this.consumerGroup,
      messageId,
    ]);
  }

  private parseMessage(id: string, fields: unknown[]): StreamMessage | null {
    try {
      const data = parseFields(fields);

      return {
        id,
        taskName: data.taskName || '',
        payload: JSON.parse(data.payload || 'null'),
        deliveryCount: 1,
        enqueuedAt: parseInt(data.enqueuedAt || '0', 10) || Date.now(),
      };
    } catch {
      return null;
    }
  }

  /**
   * Clean up ghost consumer groups from terminated workers.
   * Checks consumer idle time to identify stale groups.
   *
   * @returns Number of deleted consumer groups
   */
  async cleanup(): Promise<number> {
    let deleted = 0;

    try {
      const groups = await this.redis.send('XINFO', [
        'GROUPS',
        BROADCAST_STREAM,
      ]);
      if (!groups || !Array.isArray(groups)) return deleted;

      for (const group of groups) {
        const info = this.parseRedisInfo(group);
        const groupName = info.name as string;

        // Never delete our own group
        if (groupName === this.consumerGroup) continue;

        // Check if all consumers in this group are idle (ghosts)
        const shouldDelete = await this.isGroupIdle(groupName);
        if (shouldDelete) {
          await this.redis.send('XGROUP', [
            'DESTROY',
            BROADCAST_STREAM,
            groupName,
          ]);
          this.logger.info(`Deleted stale consumer group: ${groupName}`);
          deleted++;
        }
      }
    } catch {
      // Skip cleanup errors
    }

    return deleted;
  }

  /**
   * Check if all consumers in a group are idle (ghosts).
   */
  private async isGroupIdle(groupName: string): Promise<boolean> {
    try {
      const consumers = await this.redis.send('XINFO', [
        'CONSUMERS',
        BROADCAST_STREAM,
        groupName,
      ]);

      if (!consumers || !Array.isArray(consumers) || consumers.length === 0) {
        // No consumers = definitely stale
        return true;
      }

      for (const consumer of consumers) {
        const info = this.parseRedisInfo(consumer);
        const idle = (info.idle as number) ?? 0;

        // If any consumer is active (not idle beyond threshold), don't delete
        if (idle < this.consumerIdleThreshold) {
          return false;
        }
      }

      // All consumers are ghosts (idle beyond threshold)
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Parse Redis XINFO key-value array into object.
   */
  private parseRedisInfo(data: unknown): Record<string, unknown> {
    const info: Record<string, unknown> = {};
    if (Array.isArray(data)) {
      for (let i = 0; i < data.length; i += 2) {
        const key = data[i];
        const value = data[i + 1];
        if (typeof key === 'string') {
          info[key] = value;
        }
      }
    }
    return info;
  }
}

File: packages/backstage-go/broadcast.go
// Package backstage broadcast implementation.
// Implements a fan-out messaging pattern where tasks are delivered to all active workers.
package backstage

import (
	"context"
	"encoding/json"
	"time"

	"github.com/redis/go-redis/v9"
)

// BroadcastStream is the Redis stream key used for broadcast messages.
const BroadcastStream = "backstage:broadcast"

// BroadcastConfig for the broadcast listener.
type BroadcastConfig struct {
	ConsumerIdleThreshold time.Duration // Threshold for ghost consumer cleanup
	BlockTimeout          time.Duration
}

// DefaultBroadcastConfig returns sensible defaults.
func DefaultBroadcastConfig() BroadcastConfig {
	return BroadcastConfig{
		ConsumerIdleThreshold: time.Hour,
		BlockTimeout:          5 * time.Second,
	}
}

// BroadcastMessage represents a message from the broadcast stream.
type BroadcastMessage struct {
	ID         string
	TaskName   string
	Payload    json.RawMessage
	EnqueuedAt int64
}

// BroadcastHandler is called for each broadcast message.
type BroadcastHandler func(ctx context.Context, msg BroadcastMessage) error

// BroadcastListener listens for broadcast messages on all workers.
// Unlike standard queues where one worker processes a message, broadcast messages
// are delivered to ALL active workers (fan-out). This is achieved by creating
// a unique consumer group for each worker instance.
type BroadcastListener struct {
	redis         *redis.Client
	consumerGroup string
	consumerID    string
	handler       BroadcastHandler
	config        BroadcastConfig
	running       bool
	logger        *Logger
}

// NewBroadcastListener creates a broadcast listener.
// Each worker gets its own consumer group to receive all messages.
func NewBroadcastListener(rdb *redis.Client, workerID string, handler BroadcastHandler, config BroadcastConfig) *BroadcastListener {
	return &BroadcastListener{
		redis:         rdb,
		consumerGroup: "broadcast-" + workerID,
		consumerID:    workerID,
		handler:       handler,
		config:        config,
		logger:        NewLogger("Broadcast"),
	}
}

// Start begins listening for broadcast messages.
// It runs a blocking loop that consumes messages from the worker's unique group.
// The method returns when the context is canceled or Stop() is called.
func (b *BroadcastListener) Start(ctx context.Context) error {
	// Create unique consumer group for this worker
	err := b.redis.XGroupCreateMkStream(ctx, BroadcastStream, b.consumerGroup, "0").Err()
	if err != nil && err.Error() != "BUSYGROUP Consumer Group name already exists" {
		return err
	}

	b.running = true

	for b.running {
		result, err := b.redis.XReadGroup(ctx, &redis.XReadGroupArgs{
			Group:    b.consumerGroup,
			Consumer: b.consumerID,
			Streams:  []string{BroadcastStream, ">"},
			Count:    10,
			Block:    b.config.BlockTimeout,
		}).Result()

		if err == redis.Nil {
			continue
		}
		if err != nil {
			if b.running {
				time.Sleep(time.Second)
			}
			continue
		}

		for _, stream := range result {
			for _, msg := range stream.Messages {
				b.handleMessage(ctx, msg)
			}
		}
	}

	return nil
}

// Stop stops the broadcast listener.
func (b *BroadcastListener) Stop() {
	b.running = false
}

func (b *BroadcastListener) handleMessage(ctx context.Context, msg redis.XMessage) {
	taskName, _ := msg.Values["taskName"].(string)
	payloadStr, _ := msg.Values["payload"].(string)
	enqueuedAtStr, _ := msg.Values["enqueuedAt"].(string)

	var enqueuedAt int64
	if enqueuedAtStr != "" {
		enqueuedAt = parseInt64(enqueuedAtStr)
	}

	bm := BroadcastMessage{
		ID:         msg.ID,
		TaskName:   taskName,
		Payload:    json.RawMessage(payloadStr),
		EnqueuedAt: enqueuedAt,
	}

	if b.handler != nil {
		if err := b.handler(ctx, bm); err != nil {
			b.logger.Error("broadcast handler error", "error", err)
			return
		}
	}

	// Acknowledge
	b.redis.XAck(ctx, BroadcastStream, b.consumerGroup, msg.ID)
}

// Cleanup removes ghost consumer groups (consumers that have been idle beyond the threshold).
// This prevents the accumulation of stale consumer groups from workers that have terminated.
// It is typically called periodically by a maintenance task.
func (b *BroadcastListener) Cleanup(ctx context.Context) (int, error) {
	deleted := 0

	groups, err := b.redis.XInfoGroups(ctx, BroadcastStream).Result()
	if err != nil {
		return 0, err
	}

	for _, group := range groups {
		// Never delete our own group
		if group.Name == b.consumerGroup {
			continue
		}

		// Check if all consumers are ghosts
		if b.isGroupIdle(ctx, group.Name) {
			err := b.redis.XGroupDestroy(ctx, BroadcastStream, group.Name).Err()
			if err == nil {
				b.logger.Info("Deleted stale consumer group", "group", group.Name)
				deleted++
			}
		}
	}

	return deleted, nil
}

// isGroupIdle checks if all consumers in a group are idle (ghosts).
func (b *BroadcastListener) isGroupIdle(ctx context.Context, groupName string) bool {
	consumers, err := b.redis.XInfoConsumers(ctx, BroadcastStream, groupName).Result()
	if err != nil {
		return false
	}

	if len(consumers) == 0 {
		// No consumers = definitely stale
		return true
	}

	for _, consumer := range consumers {
		// If any consumer is active (not idle beyond threshold), don't delete
		if consumer.Idle < b.config.ConsumerIdleThreshold {
			return false
		}
	}

	// All consumers are ghosts
	return true
}

func parseInt64(s string) int64 {
	var v int64
	json.Unmarshal([]byte(s), &v)
	return v
}
